File: medium/0036-valid-sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function (board) {
  const rows = new Map();
  const cols = new Map();
  const subBoxes = new Map();

  for (let i = 0; i < 9; i++) {
    for (let j = 0; j < 9; j++) {
      if (board[i][j] == '.') continue;

      const subBoxIndex = `${Math.floor(i / 3)},${Math.floor(j / 3)}`;
      if (
        (rows.has(i) && rows.get(i).has(board[i][j])) ||
        (cols.has(j) && cols.get(j).has(board[i][j])) ||
        (subBoxes.has(subBoxIndex) &&
          subBoxes.get(subBoxIndex).has(board[i][j]))
      ) {
        return false;
      }

      if (!rows.has(i)) rows.set(i, new Set());
      if (!cols.has(j)) cols.set(j, new Set());
      if (!subBoxes.has(subBoxIndex)) subBoxes.set(subBoxIndex, new Set());

      rows.get(i).add(board[i][j]);
      cols.get(j).add(board[i][j]);
      subBoxes.get(subBoxIndex).add(board[i][j]);
    }
  }
  return true;
};

// Time complexity: O(n)
// Space complexity: O(n)
// where n is the number of cells in the grid (9x9 in this case)

File: medium/0036-valid-sudoku.py
from typing import List
from collections import defaultdict


# One pass hash set solution
# Time complexity: O(n^2)
# Space complexity: O(n^2)
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = defaultdict(set)
        cols = defaultdict(set)
        squares = defaultdict(set)

        for r in range(9):
            for c in range(9):
                if board[r][c] == ".":
                    continue

                if (
                    board[r][c] in rows[r]
                    or board[r][c] in cols[c]
                    or board[r][c] in squares[(r // 3, c // 3)]
                ):
                    return False

                rows[r].add(board[r][c])
                cols[c].add(board[r][c])

                # Tuples are immutable, so we can use them as keys
                squares[(r // 3, c // 3)].add(board[r][c])

        return True


# Time complexity: O(n^2)
# Space complexity: O(n)
class Solution2:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        # Create "seen" set to store numbers we've seen
        seen = set()

        # Check each of the rows for duplicates
        for i in range(9):
            for j in range(9):
                if board[i][j] == ".":
                    continue
                if board[i][j] in seen:
                    return False
                seen.add(board[i][j])
            seen.clear()  # Reset set for next row

        # Check each of the cols for duplicates
        for j in range(9):
            for i in range(9):
                if board[i][j] == ".":
                    continue
                if board[i][j] in seen:
                    return False
                seen.add(board[i][j])
            seen.clear()  # Reset set for next col

        # Check each of the sub-boxes for duplicates
        r = 0
        c = 0
        while r < 9 and c < 9:
            for _ in range(3):
                for i in range(r, r + 3):
                    for j in range(c, c + 3):
                        if board[i][j] == ".":
                            continue
                        if board[i][j] in seen:
                            return False
                        seen.add(board[i][j])
                # Move to next sub-box
                seen.clear()  # Reset set for next sub-box
                c += 3
            c = 0
            r += 3

        return True
